modules/src/erc721/core.rs

```diff
@@ -0,0 +1,142 @@
+    pub operator_approvals: Mapping<(Address, Address), bool>
+}
+
+impl Erc721 for Core {
+    fn balance_of(&self, owner: Address) -> U256 {
+        if owner.is_zero() {
+            revert(Error::ZeroAddress);
+        }
+        self.balances.get_or_default(&owner)
+    }
+
+    fn owner_of(&self, token_id: U256) -> Address {
+        self.owners
+            .get(&token_id)
+            .unwrap_or_revert_with(Error::InvalidTokenId)
+            .unwrap_or_revert_with(Error::InvalidTokenId)
+    }
+
+    fn safe_transfer_from(&mut self, from: Address, to: Address, token_id: U256) {
+        if !self.is_approved_or_owner(caller(), token_id) {
+            revert(Error::NotAnOwnerOrApproved);
+        }
+        self.safe_transfer(from, to, token_id);
+    }
+
+    fn transfer_from(&mut self, from: Address, to: Address, token_id: U256) {
+        if !self.is_approved_or_owner(caller(), token_id) {
+            revert(Error::NotAnOwnerOrApproved);
+        }
+        self.transfer(from, to, token_id);
+    }
+
+    fn approve(&mut self, approved: Address, token_id: U256) {
+        let owner = self.owner_of(token_id);
+        let caller = caller();
+
+        if owner == approved {
+            revert(Error::ApprovalToCurrentOwner);
+        }
+
+        if caller != owner && !self.is_approved_for_all(owner, caller) {
+            revert(Error::NotAnOwnerOrApproved);
+        }
+
+        if approved.is_zero() {
+            self.token_approvals.set(&token_id, None);
+        } else {
+            self.token_approvals.set(&token_id, Some(approved));
+        }
+
+        Approval {
+            owner,
+            approved,
+            token_id
+        }
+        .emit();
+    }
+
+    fn set_approval_for_all(&mut self, operator: Address, approved: bool) {
+        let caller = caller();
+        if caller == operator {
+            revert(Error::ApproveToCaller)
+        }
+
+        self.operator_approvals.set(&(caller, operator), approved);
+        ApprovalForAll {
+            owner: caller,
+            operator,
+            approved
+        }
+        .emit();
+    }
+
+    fn get_approved(&self, token_id: U256) -> Option<Address> {
+        self.assert_exists(&token_id);
+        self.token_approvals.get(&token_id).unwrap_or_default()
+    }
+
+    fn is_approved_for_all(&self, owner: Address, operator: Address) -> bool {
+        self.operator_approvals
+            .get(&(owner, operator))
+            .unwrap_or(false)
+    }
+}
+
+impl Core {
+    pub fn is_approved_or_owner(&self, spender: Address, token_id: U256) -> bool {
+        let owner = self.owner_of(token_id);
+        (spender == owner)
+            || self.get_approved(token_id) == Some(spender)
+            || self.is_approved_for_all(owner, spender)
+    }
+
+    fn safe_transfer(&mut self, from: Address, to: Address, token_id: U256) {
+        self.transfer(from, to, token_id);
+        // TODO: Make it safe
```

Comment:
You have to implement `ERC721Reviever` contract. @kpob already did that in DAO https://github.com/make-software/dao-contracts/blob/master/dao-erc721/src/receiver.rs

Reply:
Yup, that's why there is a todo